### ISSUE TITLE
Add support of resource limits for sriov use cases

### ISSUE THREAD

**Issue by pperiyasamy** [networkservicemesh/cmd-admission-webhook-k8s#17]
add support for sriovToken label in webhook. in case of sriov use case, service request happens with sriovToken label, in this case webhook has to inject resource specified in the label so that init container can create service request with right token id for VF attachment.

For example, pod annotation is added like this 

```
annotations:
    networkservicemesh.io: kernel://icmp-responder/nsm-1?sriovToken=worker.domain/10G
```

**Comment by denis-tingaikin** [comment]
This scenario should be worked correctly. 
Did you see any issues with using the proposed annotation?

**Comment by pperiyasamy** [comment]
@denis-tingaikin the annotation looks fine and webhook also adds it into container env variable. this works fine for example kernel mechanism veth pair use case , but for sriov, pod resource requests/limits also to be made like [here](https://github.com/networkservicemesh/deployments-k8s/tree/main/examples/use-cases/SriovKernel2Noop#run), look at patch-nsc.yaml example, isn't resource injection missing in [webhook](https://github.com/networkservicemesh/cmd-admission-webhook-k8s/blob/main/main.go#L195-L199) or I missed something to look at ?

**Comment by denis-tingaikin** [comment]
So actually, we need to add support for 
```
          resources:
            limits:
              worker.domain/10G: 1
```
Am I get it correct?

**Comment by pperiyasamy** [comment]
yes @denis-tingaikin, correct.

**Comment by denis-tingaikin** [comment]
OK, then it makes sense for me.

Can we rename the ticket to `Add support of resource limits for sriov use cases or something similar?

**Comment by pperiyasamy** [comment]
> OK, then it makes sense for me.
> 
> Can we rename the ticket to `Add support of resource limits for sriov use cases` or something similar?

sure, done, sorry if that confused you so much !

**Comment by denis-tingaikin** [comment]
@pperiyasamy all fine, thanks for clarifications!

So then to fix it we can:
1. parse  incoming annotation value
2.  check do we have `sriovToken` label.
3. If we have it then we need to add a patch with resource limits equal to `sriovToken` label value.

**Comment by mardim91** [comment]
Hello,

I intend to work on this issue. Can somebody assign me to it ?

@denis-tingaikin Can you please assign me to this issue ?

Thanks in advance

**Comment by mardim91** [comment]
I have started investigating this enhancement and the result is that is a bit more complicated than it looks. 

The approach to just add the resources to the init-container that is injected from the admission controller will not work due to the fact that the admission controller injects not only the init-container (cmd-nsc-init) but also a regular container (cmd-nsc) during the mutation process [0]. Both those containers initiate NetworkService connection requests by using client chains [1]. The problem is that the sriovtoken client in the cmd-nsc chain will not initiated properly because the environmental variable with prefix "NSM_SRIOV_TOKENS_" is empty [2]. The reason behind that is that the device plugin has filled this environmental variable only in the cmd-nsc-init container because that is the container with the injected resources. 

To solve this problem we need to share the information that the device plugin has created through the  "NSM_SRIOV_TOKENS_" also to the cmd-nsc container.

@edwarnicke , @denis-tingaikin, @fkautz  @pperiyasamy  What do you think about this problem ? Can the device plugin write the UUID token value in a shared directory among containers in a pod instead of writing it in the "NSM_SRIOV_TOKENS_" env var ?

[0]: https://github.com/networkservicemesh/deployments-k8s/blob/main/apps/admission-webhook-k8s/admission-webhook.yaml#L42-L45
      https://github.com/networkservicemesh/cmd-admission-webhook-k8s/blob/main/main.go#L77-L78

[1]: https://github.com/networkservicemesh/cmd-nsc/blob/main/main.go#L133
      https://github.com/networkservicemesh/cmd-nsc-init/blob/main/main.go#L153

[2]: https://github.com/networkservicemesh/sdk-sriov/blob/main/pkg/tools/tokens/tokens.go#L26

**Comment by mardim91** [comment]
After investigating what I have described above a bit further we have come with a solution that I will describe below:

Instead the deviceplugin to inject only the TokenIDs to the NSM_SRIOV_TOKENS_ env var will also create a file that will store that information in the host path and then will mount this file to a shared folder (EmptyDir shared volume) inside the pod's cmd-nsc-init and cmd-nsc containers.

As a next step the sriovtoken client will be enhanced properly to be able to not only read the TokenIDs only from the NSM_SRIOV_TOKENS_ env var but also from the file that has been injected from the deviceplugin in the shared volume inside the containers.

**Comment by pperiyasamy** [comment]
can you also consider skipping init container and inject only cmd-nsc container (with token resource) for sriov case ?

**Comment by mardim91** [comment]
PR for solving this issue: https://github.com/networkservicemesh/cmd-admission-webhook-k8s/pull/21

**Comment by mardim91** [comment]
@denis-tingaikin  I think we can close this issue the PR has been merged

**Comment by denis-tingaikin** [comment]
Fixed by #21.
